requests-for-comments/2022-01-27-future-of-person-model.md

```diff
@@ -0,0 +1,108 @@
+Ability to support the largest tiers of customer now and in the future
+Standard analysis such as events from paying users over time will stop yielding misleading results when people start / stop paying (as they do today)
+
+
+#### 1. After the first session we will no longer merge events prior to login with the main person’s events.
+
+![image](https://user-images.githubusercontent.com/85295485/151368075-a7ab9dc7-ee8d-4cad-a454-2524161973f1.png)
+
+How we'll map persons to events going forward:
+1. On initial visit, users device gets assigned an an (anonymous) distinct id `anonid1`. Posthog internally associates person_id `p1` to `anonid1`, which is saved together with their events
+2. When after signing up `posthog.identify('userid')` is called, we associate the same person_id `p1` with the new distinct id `userid`
+3. If the user later visits the site with a new device, a new anonymous distinct id `anonid2` is created. Posthog internally associates these events with a new person id `p2` as no mapping for that id exists.
+4. If after signing in `posthos.identify('userid')` is called, we associate events going forward with person_id `p1`
+
+As a result, events done between steps 1, 2 and 4 will be tracked as from person `p1` but events between steps 3 and 4 will be under person`p2`. As a result no events are lost but some will be considered as coming from a different (anonymous) person. In the above image this corresponds to the green period.
+
+This affects use cases where users access their product (Logged out) on multiple devices regularly. However, based on our analysis the impact is only on ~0.03% events, some marketing use-cases are more significantly affected.
+
+As a mitigation, if necessary, we would consider a buffer mechanism to reduce affected events to ~0.003%, we can provide further details on this solution, however we would ideally avoid implementing this as it increases the complexity of the system quite significantly.
+
+#### 2. Person properties become immutable: Changing properties applies only on new events and past data is not updated.
+
+Example: Count of active users broken down by `paying` user property each month in 2021.
+
+After this change, if `paying` property of a user in Dec 2021 is changed to `false`, count of `paying` users does not change for e.g. March 2021. The  user will be considered non-paying until their `paying` property changes again2.
+
+It also means users could be counted as both `paying=true` and `paying=false` during a given month when they change state.
+
+User properties will change over time on events, some customers may rely on the existing behavior (where the latest property is assumed for all prior events), however the new behavior is considered much more accurate by our customer feedback.
+
+Our proposed mitigation or most use-cases (e.g. events for users who have paid at some point) cohorts can be used to achieve the same results, we will document this for affected users
```

Comment:
> Were there any specific properties/examples for when latest property was desired?

From speaking to customers: not really. Changing properties like "email" is not a real usecase that affects day-to-day usage of the product and in practice point-in-time properties hits real usecases (like e.g. payment breakdown) much better.

That said if you have counter-examples from your usage of posthog 100% worth posting them here.

Reply:
My specific examples for email:
* needing emails to know who's still self-hosting posthog postgres for the marketing team to reach out (definitely wanted the latest emails only). Though I think here I was using some event property, but I can see others having this use case.
* there was a problem with in the app, that caused something weird to happen for certain users - we identify an event how to find those users (e.g. users running a broken version of posthog in self hosted that can end up in a crash loop or potentially set some things in a bad way and requires manual action from the instance operator) - same here I'd want latest emails.

Agree that they are likely rare and that point-in-time properties are usually what users want. Just thought it might be helpful for us to have a short snippet in the docs for everyone for how to get latest property values for anything (i.e. cohort & then potentially random insight on that cohort)

Reply:
Getting latest values definitely makes sense and is needed to even power the Persons page! 

Let's explore this more during technical planning - as you mentioned elsewhere, this need will come up elsewhere as well (e.g. calculating cohorts).